mlx/transforms.cpp

```diff
@@ -38,7 +38,13 @@ class Synchronizer : public Primitive {
 // are currently under a function transformation.
 int detail::InTracing::tracing_counter{0};
 
-void eval(std::vector<array> outputs) {
+std::shared_future<void> async_eval(std::vector<array> outputs) {
+  static std::shared_future<void> global_synchronizer;
+  // Catch up with previous async eval if needed
+  if (global_synchronizer.valid()) {
+    global_synchronizer.wait();
+  }
```

Comment:
@angeloskath do you prefer it with something like that? It's definitely safer this way.. maybe good to include it for now and revisit if it limits us.

Reply:
Yes! It looks great! The example I was using locally was

```python
x = mx.random.normal(shape=(100, 100, 100))
mx.eval(x)
for i in range(10):
    y = 2 * x
    h = mx.async_eval(y)
    z = 2 * y
    t = mx.async_eval(z)
    print(y)
```

it segfaults without the above global synchronizer.